### ISSUE TITLE
Support dynamic/overridable selectionUI in textDocument/references

### ISSUE THREAD

**Issue by matt-snider** [autozimu/LanguageClient-neovim#1002]
## Is your feature request related to a problem? Please describe.

The selection UI can only be globally configured but for some operations it would be convenient to be able to use a combination of different UIs.

In particular I have two use cases in mind for `textDocument/references`:

* Jump to a single reference (fzf):

    In this case I just want to see all references and pick one of them, for example because I'm looking for something in particular.

* Explore all references (Quickfix/LocationList):

    In this case I will be looking through many or all of the references, for example because I'm reading through the codebase and trying to understand it.

## Describe alternatives you've considered

I initially examined the code for [textDocument_references][textDocument_references] to see if it accepts
any arguments which might allow this already and concluded that it doesn't.

Then I attempted the following workaround, which doesn't work because the settings seem to be synchronized only at startup:

```viml
" Select a single reference
noremap <leader>sr :call LanguageClient#textDocument_references()<cr>

" Look through all references
noremap <leader>lr :call <SID>list_references()<cr>

function! s:list_references() abort
  " Backup and change selectionUI
  let selectionUI = g:LanguageClient_selectionUI
  let g:LanguageClient_selectionUI = "Quickfix"

  " Call method and restore
  call LanguageClient#textDocument_references()
  let g:LanguageClient_selectionUI = selectionUI
endfunction
```

[textDocument_references]: https://github.com/autozimu/LanguageClient-neovim/blob/8a2396d614579f4b17b33ec942037936b3a73625/src/language_server_protocol.rs#L1723

## Describe the solution you'd like

Ideally `LanguageClient#textDocument_references()` would accept an argument allowing the default (`g:LanguageClient_selectionUI`) to be overidden.

For example:

```viml
" select a single reference - (s)elect (r)eference
noremap <leader>sr :call LanguageClient#textDocument_references()<cr>

" list all references - (l)ist (r)eferences
noremap <leader>lr :call LanguageClient#textDocument_references({'selectionUI': 'Quickfix'})<cr>
```

## Additional context

If you think this seems like a good idea, I could try implementing it, since it seems like it would be a pretty small change.

**Comment by lrworth** [comment]
I want this for exactly the same reason!

**Comment by runiq** [comment]
This could be useful for more than just references.

I'd love to use [vim-clap](https://github.com/liuchengxu/vim-clap) as my selection UI. However, FZF is hardcoded in LCN, and I had to write my own stub implementation of `fzf#run` and `fzf#wrap` for vim-clap in order to make that work.

From what I can gather, the easiest way forward for the LCN project to allow this feature would be to expose the functions that gather the input to the selection UI. That way it'd be possible to easily write our own selection UI. The FZF integration could be re-written on top of these functions.

From what I can tell, there aren't all that many things that would need to be exposed. In fact, the context menu items are already done and could be used as a blueprint for the other sources:

- `LanguageClient#textDocument_references()` (this issue)
- `LanguageClient#findLocations()`
- `LanguageClient#textDocument_documentSymbol()`
- `LanguageClient#workspace_symbol()`
- `LanguageClient#handleCodeLensAction()`
- `LanguageClient#textDocument_codeAction()`

@autozimu What do you think? Would you be interested in a PR that attempted this?

Edit: I just realized there's also a denite source, so that'd mean that a way to get at the selection UI input already exists. Gonna go take a look over the weekend.

**Comment by autozimu** [comment]
Sounds good to me

**Comment by martskins** [comment]
Fixed by #1059
